fpnge.cc

```diff
@@ -251,38 +257,90 @@ struct HuffmanTable {
     }
   }
 
-  HuffmanTable(const uint64_t *collected_data) {
-    ComputeNBits(collected_data);
-    uint16_t bits[286];
-    ComputeCanonicalCode(nbits, bits);
+  void FillNBits() {
     for (size_t i = 0; i < 16; i++) {
       first16_nbits[i] = nbits[i];
-      first16_bits[i] = bits[i];
+      last16_nbits[i] = nbits[240 + i];
+    }
+    mid_nbits = nbits[16];
+    for (size_t i = 16; i < 240; i++) {
+      assert(nbits[i] == mid_nbits);
+    }
+    // Construct lz77 lookup tables.
+    for (size_t i = 0; i < 29; i++) {
+      for (size_t j = 0; j < (1U << kLZ77NBits[i]); j++) {
+        lz77_length_nbits[kLZ77Base[i] + j] = nbits[257 + i] + kLZ77NBits[i];
+        lz77_length_sym[kLZ77Base[i] + j] = 257 + i;
+      }
     }
+
+    dist_nbits = 1;
+
+#ifdef FPNGE_APPROX_PREDICTOR
+    approx_nbits[0] =
+        nbits[0] - 1; // subtract 1 as a fudge for catering for RLE
```

Comment:
Have you tried subtracting 1 iff nbits[0] <= (1/2)?

Even better I think could be to double all the bit costs (as that changes nothing), but remap bitcost 1 into 1 instead of 2 for 0, and perhaps bitcost 2 into 3 (or 4), or something like that.

Rationale being that RLE only kicks in for *many* consecutive 0s, so you'd need the 0 probability to be quite high.

Reply:
I haven't tried any of those. You are of course welcome to do so however.